pkg/core/resources/apis/meshidentity/providers/manager_test.go

```diff
@@ -0,0 +1,249 @@
+package providers_test
+
+import (
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"github.com/kumahq/kuma/api/common/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core"
+	"github.com/kumahq/kuma/pkg/core/kri"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	meshidentity_api "github.com/kumahq/kuma/pkg/core/resources/apis/meshidentity/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core/resources/apis/meshidentity/providers"
+	"github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/core/xds"
+	"github.com/kumahq/kuma/pkg/events"
+	core_metrics "github.com/kumahq/kuma/pkg/metrics"
+	"github.com/kumahq/kuma/pkg/test/resources/builders"
+)
+
+var _ = Describe("MeshIdentity providers", func() {
+	var eventBus events.EventBus
+	var noIdentityManager providers.IdentityProviderManager
+	var identityManager providers.IdentityProviderManager
+	now := time.Now()
+
+	BeforeEach(func() {
+		core.Now = func() time.Time {
+			return now
+		}
+		metrics, err := core_metrics.NewMetrics("")
+		Expect(err).ToNot(HaveOccurred())
+
+		eventBus, err = events.NewEventBus(10, metrics)
+		Expect(err).ToNot(HaveOccurred())
+
+		identityManager = providers.NewIdentityProviderManager(
+			providers.IdentityProviders{
+				"Bundled": &staticIdentityProvider{},
+			}, eventBus,
+		)
+
+		noIdentityManager = providers.NewIdentityProviderManager(
+			providers.IdentityProviders{
+				"Bundled": &noIdentityProvider{},
+			}, eventBus,
+		)
+	})
+
+	AfterEach(func() {
+		core.Now = time.Now
+	})
+	type testCase struct {
+		dpp                  *core_mesh.DataplaneResource
+		meshIdentities       []*meshidentity_api.MeshIdentityResource
+		expectedIdentityName string
+	}
+	DescribeTable("SelectIdentity",
+		func(given testCase) {
+			// when
+			identity := identityManager.SelectedIdentity(given.dpp, given.meshIdentities)
+
+			// then
+			Expect(identity.Meta.GetName()).To(Equal(given.expectedIdentityName))
+		},
+		Entry("select the most specific identity", testCase{
+			dpp: builders.Dataplane().WithLabels(map[string]string{
+				"app": "test-app",
+			}).AddInboundHttpOfService("test-app").Build(),
+			meshIdentities: []*meshidentity_api.MeshIdentityResource{
+				builders.MeshIdentity().WithName("not-matching-1").WithSelector(&v1alpha1.LabelSelector{
+					MatchLabels: &map[string]string{
+						"app":     "test-app",
+						"version": "v1",
+					},
+				}).Build(),
+				builders.MeshIdentity().WithName("matching-all").WithSelector(&v1alpha1.LabelSelector{
+					MatchLabels: &map[string]string{},
+				}).Build(),
+				builders.MeshIdentity().WithName("matching-specific").WithSelector(&v1alpha1.LabelSelector{
+					MatchLabels: &map[string]string{
+						"app": "test-app",
+					},
+				}).Build(),
+			},
+			expectedIdentityName: "matching-specific",
+		}),
+		Entry("select the most specific identity with 2 tags", testCase{
```

Comment:
is this test different than:

https://github.com/kumahq/kuma/blob/73c93eb34f1f821738b899758e691f8601b135dd/pkg/core/resources/apis/meshidentity/api/v1alpha1/helpers_test.go#L52-L74

also should there be a test case where the dpp has 3 labels like:

```
dpp: builders.Dataplane().WithLabels(map[string]string{
				"app":     "test-app",
				"version": "v1",
				"env": "prod",
			}).AddInboundHttpOfService("test-app").Build(),
```

and there is MID with:

```
					MatchLabels: &map[string]string{
						"app":     "test-app",
						"version": "v1",
					},
```

and another with:

```
					MatchLabels: &map[string]string{
						"app":     "test-app",
						"env":     "prod",
					},
```

which one would match then? 🤔

Reply:
It's the same test, I will remove this and add to helpers the case you mention